src/forms/Slushbucket.tsx

```diff
@@ -0,0 +1,237 @@
+
+import MDSpinner from "react-md-spinner";
+import Scrollbars from "react-custom-scrollbars";
+import ClearButton from "../components/ClearButton.js";
+
+import styles from "./SlushBucket.css";
+import formStyles from "../styles/Forms.css";
+import inputStyles from "../styles/Input.css";
+import displayStyles from "../styles/Display.css";
+
+type choiceT = {value: string | number, display: string};
+export type choicesT = [choiceT];
+type valueItemT = string | number
+type valueT = [valueItemT];
+
+interface Props {
+  title: string,
+  name: string,
+  placeholder: string,
+  readonly: boolean | undefined,
+  isFetching: boolean | undefined,
+  value: valueT,
+  choices: choicesT,
+  validators?: Function[],
+  validated: boolean,
+  valueChanged: Function,
+};
+
+interface State {
+  searchString: string
+}
+
+export default class SlushBucket extends Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      searchString: ""
+    };
+    this.handleKeyUp = this.handleKeyUp.bind(this);
+  }
+
+  handleKeyUp(e: any): void {
+    if (e.key === "Escape") this.resetSearchString();
+  }
+
+  resetSearchString() {
+    this.setState({ searchString: "" });
+  }
+
+  handleInput(e: any): void {
+    this.setState({ searchString: e.target.value });
+  }
+
+  render() {
+    const {
+      choices,
+      value,
+      isFetching,
+      placeholder,
+      valueChanged
+    } = this.props;
+
+    const selected = value;
+    const updateModelValue = valueChanged;
+
+    return (
+      <div className={`${styles.SelectChoice} form-input`}>
+        <div
+          style={{
+            width: "50%",
+            display: "flex",
+            justifyContent: "space-between"
+          }}
+        >
+          <input
+            tabIndex={-1}
+            type="text"
+            autoComplete="false"
+            className={
+              formStyles.FormControl +
+              " " +
+              (this.props.readonly ? inputStyles.ReadOnly : "")
+            }
+            placeholder={placeholder}
+            onChange={e => this.handleInput(e)}
+            onKeyUp={event => {
+              this.handleKeyUp(event);
+            }}
+            value={this.state.searchString}
+            disabled={this.props.readonly}
+            readOnly={this.props.readonly}
+          />
+          <div
+            className={`${styles.Spinner} ${isFetching
+              ? displayStyles.Block
+              : displayStyles.None}`}
+          >
+            <MDSpinner size={18} />
+          </div>
+
+          <div
+            style={{
+              transform: "translateX(-33px)",
+              display: "flex",
+              alignItems: "center",
+              visibility: this.state.searchString === "" ? "hidden" : "visible"
+            }}
+          >
+            <ClearButton
+              onClick={() => {
+                this.resetSearchString();
+              }}
+            />
```

Comment:
If we also put .bind(this) for this one in the constructor, then this can just be onClick={resetSearchString}.